lib/hash.c

```diff
@@ -276,24 +276,34 @@ M0_INTERNAL uint64_t m0_htable_size(const struct m0_htable *htable)
 	return len;
 }
 
+/*
+ * Improved version of well-known djb2 hash-function.
+ *
+ * Here is an example of what it does with a big number
+ * which changes only slightly from the beginning or end:
+ *
+ * hash 8aee7dc0513403da -> 612cbbc11beedeed
+ * hash 7aee7dc0513403da -> 7f5a5f634d50c0ba
+ * hash 7aee7dc0513403db -> af729f8afb4348f8
+ */
+static uint64_t hash(void *buf, int n)
+{
+	uint64_t hash = 5381;
+	unsigned char *p;
+
+	for (p = buf; n-- > 0; p++) {
+		hash += ((unsigned long long)*p << 32) | *p;
+		hash *= 2971215073;
+		hash >>= 16;
+		hash *= 2971215073;
+	}
+
+	return hash;
+}
```

Comment:
Ah, polynomial hashes. It's easy to check that the value of hash(p, n) is (modulo 2^64) `p[0]*a*b^n + p[1]*a*b^{n-1} + ... p[n-1]*a*b` for certain a and b. Therefore, if you have some string R, such that `hash(R) == 0`, then for any string X, `hash(R + X) == hash(R + R + X) == ... == 0`, where + is string catenation. That's not very secure.

Reply:
If we need security here, we should probably use SHA-256 or something.

Reply:
Please make sure this piece of code is not just copied from somewhere.
This might violate the copyright.

Reply:
@huanghua78 It is not. The only line which is copied (from http://www.cse.yorku.ca/~oz/hash.html) is `uint64_t hash = 5381;` and even it has changed type. Anyway, it seems the original algorithm is in public domain, so I think we are safe.

Reply:
OK. Thank you.